doc: Add introduction of launching Celadon User OS (#3324)
projectacrn/acrn-hypervisor: doc/tutorials/using_celadon_as_uos.rst

      +--------------------------------------------------------------------+-------------------------------------------+

   .. note:: You may need to modify the source code manually if meet the error when use
      ``git apply`` command to apply the patches above.

#. Build Celadon image::

   $ device/intel/mixins/mixin-update
   $ make SPARSE_IMG=true gptimage -j $(nproc)

.. note:: Replace the ``$(nproc)`` argument with the number of processor threads on your workstation
   in order to build the source code with parallel tasks. The Celadon gptimage will be
   generated to ``out/target/product/cel_apl/cel_apl_gptimage.img``

Steps for Using Celadon as User OS
**********************************

#. Follow :ref:`getting-started-apl-nuc` to boot the "ACRN Service OS"

#. Download ``OVMF.fd`` from `link
   <https://raw.githubusercontent.com/projectacrn/acrn-hypervisor/master/doc/tutorials/OVMF.fd>`__

#. Download ``launch_android.sh`` script from `link
   <https://raw.githubusercontent.com/tanminger/acrn-hypervisor/ed7a758377151b4c7296c0b5669167585b50907b/devicemodel/samples/nuc/launch_android.sh>`__

> #. Boot Celadon as User OS::

Reviewer: Change to:
```
#. Use ``scp`` to copy these downloaded files from your development system to your NUC::
```
Is there a reason you can't download the files in the previous step  on the NUC directly instead of copying them from your development system?

Author: I was planing to let user download directly to the NUC from the repo in the steps, but the link is too long and looks strange, so I give up :smile:

Reviewer: Well, it's the same link you have to use on the development system, or are you thinking you'd click the link in the documentation on your development system, but you'd have to type it in on your NUC?  I'm not sure what the issues is?

Author: Now it's verified that Celadon UOS is able to boot successfully with the build-in ``OVMF.fd`` and ``launch_win.sh`` script, I will update the steps in next release.

   # mkdir ~/celadon && cd ~/celadon
   # scp <OVMF.fd from your host> ./
   # scp <cel_apl_gptimage.img from your host> ./android.img
   # scp <launch_android.sh from your host> ./
   # sh launch_android.sh

> #. You will see the shell console after boot Celadon, and check build info with ``getprop`` command:

Reviewer: ```suggestion
#. You will see the shell console after Celadon boots. Check the build info using the ``getprop`` command:
```

.. _using_celadon_as_uos:

Using Celadon as User OS
########################

`Celadon <https://01.org/projectceladon/>`_ is an open source Android* software reference stack
for Intel architecture. It builds upon a vanilla Android stack and incorporates open sourced components
that are optimized for the hardware. This tutorial describes how to run Celadon as the User OS
on the ACRN hypervisor. We are using Kaby Lake-based NUC (model NUC7i7DNHE) in this tutorial.
Other :ref:`ACRN supported platforms <hardware>` should work as well.

> Introduction to OVMF

Reviewer 2: I don't think it makes much sense to introduce OVMF here since we're only going to use a pre-compiled binary. I would get rid of this paragraph altogether.

Author: Sure, I will remove this part since the built-in OVMF is also work fine.

********************

Open Virtual Machine Firmware (OVMF) is an `EDK II
<https://github.com/tianocore/tianocore.github.io/wiki/EDK-II>`_
based project to enable `UEFI <https://github.com/tianocore/tianocore.github.io/wiki/UEFI>`_
support for Virtual Machines. OVMF contains sample UEFI firmware for `QEMU <http://www.qemu-project.org/>`_ and `KVM
<http://www.linux-kvm.org/>`_.

Prerequisites
*************

> * Ubuntu 16.04 or higher with at least 150G free disk space

Reviewer 2: We would like to get rid of Ubuntu 16.04 (there are too many problems if you're trying to build ACRN on it), can we refer to Ubuntu 18.04 or higher instead? Is it known to work for a fact?

Reviewer 2: Thanks @ttzeng ! @lirui34 , can you update this tutorial to only reference Ubuntu 18.04 and above?

> * Intel NUC Kit ( Reference Platforms: :ref:`ACRN supported platforms <hardware>` )

Author: Yes, I've checked ``device mode`` option is also exit in the APL NUC Bios setting.

> * Intel NUC Kit ( Reference Platforms: :ref:`ACRN supported platforms <hardware>` )

Reviewer 3: @lirui34 It looks like the USB Device mode is supported since firmware 0059 for NUC6CAYH, and firmware NUC7ixDNH. Suggest to call call it out explicitly. :)

> * Intel NUC Kit ( Reference Platforms: :ref:`ACRN supported platforms <hardware>` )

Author: Thanks for the information, I will suggest people to use 0059 or above so that they will also see the ``device mode`` in APL BIOS settings.
Unfortunately in my latest test result based on NUC7i7BNH and NUC6CAYH, there's no Celadon graphic on those platforms with only one HDMI port. Since I am not sure it's a Celadon issue or ACRN issue, so I don't want to mention the APL in this introduction.

* BIOS version should be 0057 or later, the ``Device Mode`` option is selected
  for the internal USB 3.0 port header on the USB category of the Devices tab
* If you need to use ``adb`` or ``fastboot`` tool with a Type-A to Type-A USB cable
  for the Celadon User VM, then you must provide a USB dongle (e.g. `Dawson Canyon USB 3.0 female
  to 10-pin header cable <https://www.gorite.com/dawson-canyon-usb-3-0-female-to-10-pin-header-cable>`_)
  to connect to the internal USB 3.0 port header on the main board of the Intel NUC system.
  Refer to the `Technical Product Specification
  <https://www.intel.com/content/dam/support/us/en/documents/mini-pcs/nuc-kits/NUC7i5DN_TechProdSpec.pdf>`_
  to identify the USB 3.0 port header on the main board.
* Two HDMI monitors

Build Celadon from source
*************************

#. Follow the `Build Celadon from source instructions
   <https://01.org/projectceladon/documentation/getting_started/build-source>`_
   to get the Celadon project source code ready.

#. Select Celadon build target::

      $ cd <Celadon project directory>
      $ source build/envsetup.sh
      $ lunch cel_apl-userdebug

> 

Author: I met several troubles to build the ``cel_kbl`` target, while I did not see them with the ``cel_apl`` target (Actually a ``cel_apl`` img is also works on a KBL NUC).  So I would like to use the apl target instead in order to get all the process smoothly.

> 

Reviewer 3: Then I'd suggest to change the NUC kit model to NUC6CAYH, instead of NUC7i7DNH

> 

Author: As the reason above, I will still use NUC7i7DNH in this introduction. :-)

   .. note:: You can run ``lunch`` with no arguments to manually choose your Celadon build variants.

> #. Download these additional patches and apply each one individually with the command::

Reviewer 2: When do you expect these to be merged?

It may be good to indicate which revision this tutorial is based on (and which needs all these patches). So the reader can more easily track down if a newer version has it already or not.

Reviewer 2: OK, this is good to know. Thanks @ttzeng

      | Patch link                                                         | Description                               |
      +====================================================================+===========================================+
      | https://github.com/projectceladon/device-androidia/pull/458        | kernel config: Add the support of ACRN    |
      +--------------------------------------------------------------------+-------------------------------------------+
      | https://github.com/projectceladon/device-androidia-mixins/pull/293 | graphic/mesa: Add the support of ACRN     |
      +--------------------------------------------------------------------+-------------------------------------------+
      | https://github.com/projectceladon/device-androidia/pull/441        | cel_apl: use ttyS0 instead of ttyUSB0     |
      +--------------------------------------------------------------------+-------------------------------------------+
      | https://github.com/projectceladon/device-androidia/pull/439        | Disable trusty and pstore                 |

>       +--------------------------------------------------------------------+-------------------------------------------+

Author: Thanks for the information, I will try it and see if the patch list should be update. :smile:

>       +--------------------------------------------------------------------+-------------------------------------------+

Author: I was able to boot Trusty Celadon on SOS but the ``acrn-dm`` has to be replaced since the default one is not support Trusty storage.
But so far I do not know exactly how to build such ``acrn-dm``, maybe it should be another tutorial :-)

>       +--------------------------------------------------------------------+-------------------------------------------+

Reviewer 2: > I was able to boot Trusty Celadon as UOS but the `acrn-dm` has to be replaced since the default one is not support Trusty storage.

Do you have the details behind this? There is no flag to be used when building `acrn-dm`, there is a parameter called `--enable-trusty` to be used when starting the VM. Based on your comment, it sounds like there are patches needed that should also be sent upstream?

>       +--------------------------------------------------------------------+-------------------------------------------+

Author: > Do you have the details behind this? There is no flag to be used when building `acrn-dm`, there is a parameter called `--enable-trusty` to be used when starting the VM. Based on your comment, it sounds like there are patches needed that should also be sent upstream?

Exactly, `--enable-trusty` is not functional with the built-in ``acrn-dm`` (looks like a fake parameter ?). I've got a patch from @YadongQi which is supported for NUC platforms, but it is not flexible enough to send upstream so far... maybe we should talk with @NanlinXie for the next step ?

>       +--------------------------------------------------------------------+-------------------------------------------+

Reviewer 2: Thanks for the clarification. In this case, I would suggest we merge this PR as it is and track enabling Trusty separately in a later PR. Posting a patch not suitable for upstream is likely to become obsolete rather quickly. How long would it take to refactor the patch so it's suitable to be upstreamed?

>       +--------------------------------------------------------------------+-------------------------------------------+

Reviewer 4: You may reference Celadon's doc: https://github.com/projectceladon/celadon-documentation/blob/master/source/tutorials/acrn_ovmf/acrn_ovmf_celadon.rst
BTW, the `--enable-trusty` is not a fake param, just need changes to make OVMF support trusty. And the patches are maintained by @peterfang. Peter, do you have any plan to upstream the patches.

   .. note:: If the ``git apply`` command shows an error, you may need to modify
      the source code manually instead.

#. Build Celadon image::

   $ device/intel/mixins/mixin-update
   $ make SPARSE_IMG=true gptimage -j $(nproc)

.. note:: Replace the ``$(nproc)`` argument with the number of processor threads on your workstation
   in order to build the source code with parallel tasks. The Celadon gptimage will be

>    generated to ``out/target/product/cel_apl/cel_apl_gptimage.img``

Author: I'd like to use ``cel_apl`` since it's easy to start :smile:

Steps for Using Celadon as User OS
**********************************

#. Follow :ref:`getting-started-apl-nuc` to boot the "ACRN Service OS"

#. Download the `OVMF.fd firmware image

>    <https://raw.githubusercontent.com/projectacrn/acrn-hypervisor/master/doc/tutorials/OVMF.fd>`_

Reviewer 2: Can you not use the `OVMF.fd` that's already in the repo instead? https://github.com/projectacrn/acrn-hypervisor/tree/master/devicemodel/bios

Author: Yes, I verified that ``OVMF.fd`` is also work for Celadon UOS, so I will update this part in next release.

>    from the ACRN GitHub repo to your development system.

Author: It will use ``/usr/share/acrn/bios/OVMF.fd`` in the new release

#. Download the `launch_android.sh script
   <https://raw.githubusercontent.com/tanminger/acrn-hypervisor/ed7a758377151b4c7296c0b5669167585b50907b/devicemodel/samples/nuc/launch_android.sh>`_

>    to your development system.

Reviewer 2: Why the link to a specific commit in someone's repo? I would add such file to our existing folder containing launch scripts here: https://github.com/projectacrn/acrn-hypervisor/tree/master/devicemodel/samples/nuc

Author: I will use ``launch_win.sh`` script instead since that's also functional.

Reviewer 2: I have a slight preference for adding a new script called `launch_celadon.sh` (or `launch_android.sh`), we already have various scripts in there for different OSs and adding one more will bring less confusion than having to use one meant for Windows. (not mentionning the fact that if we later modify `launch_win.sh` I'm not sure we'll always remember to test those changes with Android/Celadon).

#. Use ``scp`` to copy these downloaded files from your development system to your NUC::

   # mkdir ~/celadon && cd ~/celadon

>    # scp <OVMF.fd from your host> ./

Author: Fixed in next release

   # scp <cel_apl_gptimage.img from your host> ./android.img
   # scp <launch_android.sh from your host> ./

> #. You will see the shell console after Celadon boots. Check the build info using the ``getprop`` command:

Author: Yes, I will move the launch script operation from the step below

      +--------------------------------------------------------------------+-------------------------------------------+
      | https://github.com/projectceladon/device-androidia-mixins/pull/293 | graphic/mesa: Add the support of ACRN     |
      +--------------------------------------------------------------------+-------------------------------------------+
      | https://github.com/projectceladon/device-androidia/pull/441        | cel_apl: use ttyS0 instead of ttyUSB0     |
      +--------------------------------------------------------------------+-------------------------------------------+
      | https://github.com/projectceladon/device-androidia/pull/439        | Disable trusty and pstore                 |
      +--------------------------------------------------------------------+-------------------------------------------+

   .. note:: If the ``git apply`` command shows an error, you may need to modify
      the source code manually instead.

#. Build Celadon image::

   $ device/intel/mixins/mixin-update
   $ make SPARSE_IMG=true gptimage -j $(nproc)

> .. note:: Replace the ``$(nproc)`` argument with the number of processor threads on your workstation

Reviewer 2: You don't need to replace `$(nproc)`, it will be done automatically when you run the `make` command (try to run `echo $(nproc)` on your Linux system if you want to see it in action).

Author: I really didn't know the meaning of ``$(nproc)``, now I get it. thank you so much @gvancuts :smile:
